### ISSUE TITLE
reproducing best classifier output

### ISSUE THREAD

**Issue by erexhepa** [dsibournemouth/autoweka#7]
Hi,

I was recently using the autoweka classifer to optimize the algorithm and the hyperparameter selection. I got at the end of the run a classifier (random forest) and the corresponding optimal parameters. I tried to reproduce the accuracy and the classifications results outputed by autoweka by calling random forest with the optimal parameters outputed by autoweka and I didn't get at all the same performances using CV (10-fold) on the same dataset.

Is it possible to get a full trace of the optimal algorithm with all the parameters values ? maybe some of the default parameters when calling RF are not the same as the default autoweka param (the ones not reported by autoweka) ? 

thanks

Elton

**Comment by erexhepa** [comment]
Hi Manuel,

Thanks for your quick reply and the info above. Really useful. 

Just wanted to add with regard to my first post above that I'm using the package autoweka you can download with the package manager from the weka interface. Just wanted to clarify this as the autoweka standalone application here might be different from the package but please do not hesitate to comment if this is the case.

But with regard to your comments above, I did choose the same seed for the random generator so the fold repartation of the samples should be the same for autoweka and randomforest when calling it outside autoweka. Also I'm using the same weka distribution (meaning the same GUI) but do you mean the package autoweka has it's own weka internal distribution ?

Elton

**Comment by DraXus** [comment]
Hi Elton,

The version in this repository is an extension of Auto-WEKA 0.5, which is standalone and use internally WEKA 3.7.2. The current official version of Auto-WEKA is delivered as WEKA package and use the recent WEKA 3.8 (you can find it here https://github.com/automl/autoweka/ ), so I guess that's the one you are using it.

Take into account that there are several seeds. The one that can you set from the GUI is the one for the starting the optimisation process. The seed for the CV split is in autoweka.instances file inside your experiment folder (by default is 0) - this is the one you have to set in the "Random seed for XVal" option in the GUI ( http://i.imgur.com/1MtJeqD.png ). Also Random Forest classifier has a seed (by default is 1). There might be additional seeds depending on the methods you are using.

Sorry if I'm saying obvious things for you here, just wanted to clarify things that are not straightforward to see when you start working with Auto-WEKA.

Regards,
Manuel

**Comment by erexhepa** [comment]
Hi Manuel,

Thanks for your reply. I set the seed in several places to be the same but will double check again and will have a look at the places you mentioned above. Thanks a mill.

Elton

**Comment by DraXus** [comment]
No problem. I'll close this issue, but feel free to open if you need.

**Comment by erexhepa** [comment]
Apologies for still annoying you with this :) but maybe you can save me sometime meanwhile I'm looking on my own. I think it would be very interesting to see the details of every tested model but was wondering where is the log writen of the autoweka algorithm installed via the package manager of weka ? There is the standart weka log but there is nothing particulrly detailed there beside the smac score. Is there a default DIR ? I'm checking my home dir meanwhile.

Thanks a lot Manuel,

Elton

**Comment by DraXus** [comment]
Hi Elton,

To be honest I haven't worked with the recent version of Auto-WEKA so I don't know where are the logs being stored. In the standalone version they are in the experiment folder. Please feel free to ask in https://github.com/automl/autoweka/, I'm sure @larskotthoff can help with that.

Regards,
Manuel

**Comment by larskotthoff** [comment]
Let me start with a few general things on Auto-WEKA's generalisation performance. In every case, the reported values are *estimates* of the performance on new data -- there are no guarantees. In particular, even though Auto-WEKA takes care not to overfit, it will occur in some cases and it'll report really good performance that you can't reproduce. We're working on ways of addressing this, but for now the best you can do (and what we usually do) is run Auto-WEKA a couple of times with different random seeds. The average performance across these runs is usually a good indicator of what can be achieved in general.

Ok, so what does Auto-WEKA do internally? In the most recent versions, the data you give it is split into training and testing (roughly 2/3 and 1/3) and then the training set is further partitioned into 10 cross-validation folds. So Auto-WEKA uses both cross-validation and a validation set to ensure good performance in general (although all of these are only probabilistic guarantees, so they will fail in some cases). 

As for your actual question: Auto-WEKA creates a temporary folder at $TMPDIR/autoweka[0-9]+ for every run. In this directory, you can find lots of output and some detail on the models. Auto-WEKA doesn't save *everything* though, so what you're looking for may not be there.

**Comment by erexhepa** [comment]
yeah I was suspecting something like that as well. I did some boostrapping to study a bit the variance of the performances and you can see that the high performance is due to overfitting. Getting the right split. 

Is very interesting what you guys have done and a very useful exercise for model exploration but a detailed log somewhere would be very useful for this purpose. You can go back see in details model where fit and how where the data distributed. 

we do not have a lot of data at hand and that makes the models very unstable and having the hand on how the training/validation is done would interesting. 

I saw there is a python wrapper around sklearn. Would you suggest to use that to tune a bit more how automated model, hypermater optimisation is done ? (Trying to get a linux machine to install it at the moment) 

Thank you very much for you comments

Elton

**Comment by larskotthoff** [comment]
The detailed logging isn't done because that would produce a *lot* of output in most cases. It shouldn't be too hard to modify the code to do so though.

Auto-sklearn is a sister project that uses the same technology.

**Comment by erexhepa** [comment]
ok, yeah if it's not too much work then modifing the package is an option indeed. OK thanks.

**Comment by felipeebert** [comment]
Dear @larskotthoff,

Is the problem of reproducing the best classifier output from Auto-Weka still an issue? 

I'm running Weka 3.8.1 and the Auto-Weka installed from its package manager (version 2.5), and I can't get even close results to the best classifier Auto-Weka suggests... 

Thanks in advance,
Felipe

**Comment by larskotthoff** [comment]
As I've said above, there's no guarantee (and there never will be!) that the same performance can be achieved with different data or even with the same data split differently.
